// @vikelabs/uvic-course-scraper/dist/common/urls.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.courseDetailUrl = exports.coursesUrl = exports.subjectsUrl = exports.courseListingTermUrl = exports.courseListingsEntriesUrl = exports.detailedClassInformationUrl = exports.classScheduleListingUrl = void 0;
/* eslint-disable camelcase */
var qs = __importStar(require("querystring"));
var BASE_URL = 'https://www.uvic.ca/BAN1P';
// URL builders for known pages on UVic BAN1P.
// Also serves as some documentation for known and useful endpoints.
// GET: Class Schedule Listing
// eg. https://www.uvic.ca/BAN1P/bwckctlg.p_disp_listcrse?term_in=202009&subj_in=PAAS&crse_in=138&schd_in=
var classScheduleListingUrl = function (term, subject, course) {
    var params = qs.stringify({
        term_in: term,
        subj_in: subject,
        crse_in: course,
        schd_in: '',
    });
    return BASE_URL + "/bwckctlg.p_disp_listcrse?" + params;
};
exports.classScheduleListingUrl = classScheduleListingUrl;
// GET: Detailed Class Information
var detailedClassInformationUrl = function (term, crn) {
    var params = qs.stringify({
        term_in: term,
        crn_in: crn,
    });
    return BASE_URL + "/bwckschd.p_disp_detail_sched?" + params;
};
exports.detailedClassInformationUrl = detailedClassInformationUrl;
// GET: Course Listing Entries
// Note: can hit this endpoint with a POST request too. Difference between
// eg. https://www.uvic.ca/BAN1P/bwckctlg.p_display_courses?term_in=202009&one_subj=PAAS&sel_crse_strt=138&sel_crse_end=138&sel_subj=&sel_levl=&sel_schd=&sel_coll=&sel_divs=&sel_dept=&sel_attr=
var courseListingsEntriesUrl = function (term, subject, course) {
    // TODO: figure out what the result of the fields mean.
    var params = qs.stringify({
        term_in: term,
        one_subj: subject,
        sel_crse_strt: course,
        sel_crse_end: course,
        sel_subj: '',
        sel_levl: '',
        sel_schd: '',
        sel_coll: '',
        sel_divs: '',
        sel_dept: '',
        sel_attr: '',
    });
    return BASE_URL + "/bwckctlg.p_display_courses?" + params;
};
exports.courseListingsEntriesUrl = courseListingsEntriesUrl;
// POST: Course Listing Term
// eg. https://www.uvic.ca/BAN1P/bwckctlg.p_disp_cat_term_date
// qs: call_proc_in=bwckctlg.p_disp_dyn_ctlg&cat_term_in=202001
var courseListingTermUrl = function (term) {
    var params = qs.stringify({
        //   TODO: not sure what this does
        call_proc_in: 'bwckctlg.p_disp_dyn_ctlg',
        cat_term_in: term,
    });
    return BASE_URL + "/bwckctlg.p_disp_cat_term_date?" + params;
};
exports.courseListingTermUrl = courseListingTermUrl;
// construct subjects endpoint
function subjectsUrl(catalogId) {
    return "https://www.uvic.ca/BAN1P/pkg_kuali_api.pr_get_catalog?p_catalog=" + catalogId;
}
exports.subjectsUrl = subjectsUrl;
// UVic Legacy Calendar
// const LEGACY_BASE_URL = (year: string, month: string) => `https://web.uvic.ca/calendar${year}-${month}/CDs/`;
function coursesUrl(catalogId) {
    return "https://uvic.kuali.co/api/v1/catalog/courses/" + catalogId;
}
exports.coursesUrl = coursesUrl;
function courseDetailUrl(catalogId, pid) {
    return "https://uvic.kuali.co/api/v1/catalog/course/" + catalogId + "/" + pid;
}
exports.courseDetailUrl = courseDetailUrl;
